container/multi-value-slice/multi_value_slice.go

```diff
@@ -0,0 +1,380 @@
+package multi_value_slice
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/prysmaticlabs/prysm/v4/container/multi-value-slice/interfaces"
+)
+
+// MultiValueSlice defines an abstraction over all concrete implementations of the generic Slice.
+type MultiValueSlice[O interfaces.Identifiable] interface {
+	Len(obj O) int
+}
+
+// Value defines a single value along with one or more IDs that share this value.
+type Value[V any] struct {
+	val V
+	ids []interfaces.Id
+}
+
+// MultiValue defines a collection of Value items.
+type MultiValue[V any] struct {
+	Values []*Value[V]
+}
+
+// Slice is the main component of the multi-value slice data structure. It has two type parameters:
+//   - V comparable - the type of values stored the slice. The constraint is required
+//     because certain operations (e.g. updating, appending) have to compare values against each other.
+//   - O interfaces.Identifiable - the type of objects sharing the slice. The constraint is required
+//     because we need a way to compare objects against each other in order to know which objects
+//     values should be accessed.
+type Slice[V comparable, O interfaces.Identifiable] struct {
+	sharedItems     []V
+	individualItems map[uint64]*MultiValue[V]
+	appendedItems   []*MultiValue[V]
+	cachedLengths   map[interfaces.Id]int
+	lock            sync.RWMutex
+}
+
+// Init initializes the slice with sensible defaults. Input values are assigned to shared items.
+func (s *Slice[V, O]) Init(items []V) {
+	s.sharedItems = items
+	s.individualItems = map[interfaces.Id]*MultiValue[V]{}
+	s.appendedItems = []*MultiValue[V]{}
+	s.cachedLengths = map[interfaces.Id]int{}
+}
+
+// Len returns the number of items for the input object.
+func (s *Slice[V, O]) Len(obj O) int {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	l, ok := s.cachedLengths[obj.Id()]
+	if !ok {
+		return len(s.sharedItems)
+	}
+	return l
+}
+
+// Copy copies items between the source and destination.
+func (s *Slice[V, O]) Copy(src O, dst O) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for _, item := range s.individualItems {
+		for _, v := range item.Values {
+			found := compareIds(v.ids, src.Id(), func(_ int) {
+				v.ids = append(v.ids, dst.Id())
+			})
```

Comment:
Not sure I fully understand this function call. Seems like found is a boolean, but compareIds has a side effect of appending? Would be good to either rename or clarify with commentary

Reply:
We look for the Id of `src` and append the Id of `dst`